added page on vlan trunking (#33)
bisdn/bisdn.github.io: system-configuration/layer2/trunk-vlan.md

---
title: Trunking
parent: Layer 2
nav_order: 2
---

# Trunk VLAN

## Introduction/motivation

> A trunk port can carry traffic for all VLANs available on a switch, which is

Reviewer: What does "carry" mean? What does "available on a switch" mean?

Author: changed wording to forward and configured

> commonly named trunking. The trunk port tags incomming packets and diretcts them

Reviewer: a trunk port is commonly named trunking?

Author: missing words

> commonly named trunking. The trunk port tags incomming packets and diretcts them

Reviewer: ```suggestion
commonly named trunking. The trunk port tags incoming packets and directs them
```

> commonly named trunking. The trunk port tags incomming packets and diretcts them

Reviewer: was does "directs" mean?

Author: I guess forward is the right word

to their designated VLAN.

## Example switch with uplink

> We will give an example configuration for a switch with two access ports on a single

Reviewer: ```suggestion
The example configuration below shows a switch with two access ports on a single
```

Author: fixed

VLAN and a trunk port for the uplink connection.
![switch_uplink](system-configuration/layer2/figures/trunk.svg)

The access ports ``port1`` and ``port2`` have servers connected to them and
should be on VLAN 101. For simplicity of server configuration PVID 101, and

> untagged egress traffic will be set on these ports so the servers are not aware

Reviewer: this sentence should be completely rephrased

of the VLAN.
``port48`` is the switch uplink connection, which receives and forwards traffic

> on VLAN 101 as well. The uplink connection expects traffic to be on VLAN 101, so

Reviewer: "as well" ?

> here we will not use untagged egress traffic. 

Reviewer: the "we" should probably be dropped

Author: I don't really mind the use of personal pronouns.

> The configuration fort his example can be done with either ``iproute2`` for

Reviewer 2: ```suggestion
The configuration for this example can be done with either ``iproute2`` for
```

Author: fixed

> Configuration will first be given using ``iproute2`` and then systemd-network files, where

Reviewer: ```suggestion
The configuration fort his example can be done with either ``iproute2`` for testing or systemd-network files for persistency in production environments.
```

Author: agreed, better wording

> only the latter one is reboot persistent.

Reviewer: ```suggestion
```

Author: fixed

## iproute2

Bridge creation is done with the following command.

```
ip link add name swbridge type bridge vlan_filtering 1 vlan_default_pvid none
ip link set swbridge up
```

The primary VLAN identifier (PVID) is used to tag incoming traffic that does not
have any VLAN tag. Since we also want a trunk port connected to the bridge, we

> ned to disable it on the bridge, and add it to all ports separately. 

Reviewer: ```suggestion
need to disable it on the bridge, and add it to all ports separately. 
```

Author: fixed

> ned to disable it on the bridge, and add it to all ports separately. 

Reviewer: why?

```
# port1
ip link set port1 master swbridge 
ip link set port1 up

# port2
ip link set port2 master swbridge
ip link set port2 up

# port48
ip link set port48 master swbridge
ip link set port48 up
```

> Finally, configuring the VLANs on the bridge member ports, and the bridge itself is done with the following commands. The bridge port will inherit the VLANs and PVID from the bridge, unless specifically overwritten. The self flag is required when configuring the VLAN on the bridge interface itself.

Reviewer: what does "self" do?

Author: This is just literally what documentation says, there is no explanation.
'self   the vlan is configured on the specified physical device.
         Required if the device is the bridge device.'

Reviewer 2: From Cumulus: https://docs.cumulusnetworks.com/cumulus-linux/Layer-2/Ethernet-Bridging-VLANs/#bridge-fdb-command-output

```
bridge vlan add vid 101 dev port1 pvid
bridge vlan add vid 101 dev port2 pvid
bridge vlan add vid 101 dev port48 untagged
bridge vlan add vid 101 dev swbridge self
```

Removing the configuration can be done with a reboot, or by deleting the bridge.
```
ip link del swbridge
```

## systemd-networkd

> The configuration with systemd-networkd can be done with the following files,

Reviewer: ```suggestion
The configuration with systemd-networkd can be done by placing the following files,
```

Author: thanks fixed

> under the /etc/systemd/networkd directory. The first line of the snippet give

Reviewer: ```suggestion
under the /etc/systemd/networkd directory. The first line of the snippet is
```

Author: fixed

the file name.

> The first two files creates the bridge and adds VLAN 101 to it, completely

Reviewer: ```suggestion
The first two files create the bridge and adds VLAN 101 to it, completely
```

Author: fixed

analogous to ``iproute2``

```
10-swbridge.netdev:

[NetDev]
Name=swbridge
Kind=bridge

[Bridge]
VLANFiltering=1
DefaultPVID=none
```

```
10-swbridge.network:

[Match]
Name=swbridge

[BridgeVLAN]
VLAN=101
```

Attaching the access ports ``port1`` and ``port2`` is done as follows

```
20-port1.network:

[Match]
Name=port1

[Network]
Bridge=swbridge

[BridgeVLAN]
VLAN=101

> EgressUntagged=1

Reviewer 3: > EgressUntagged=
>    The VLAN ID specified here will be used to untag frames on egress.

You are setting VLAN ID 1 to EgressUntagged, is this what you want?

Author: indeed not, it should be 101

PVID=101
```

```
20-port2.network:

[Match]
Name=port2

[Network]
Bridge=swbridge

[BridgeVLAN]
VLAN=101
EgressUntagged=1
PVID=101
```

> Adding PVID to a port implies adding the VLAN as well, thus he line with VLAN=101 can be left out , but is kept here for clarity.

Reviewer: Are you sure about that? I am pretty sure you do not need the VLAN if your server only sends and receives untagged traffic.

Author: https://www.freedesktop.org/software/systemd/man/systemd.network.html#PVID=
Are we talking about the same thing here? I misunderstood the docs anyway I think.
If you add PVID, systemd-network says that this implies that the port also must have that VLAN configured.

This port is on the switch, not on the server. Or am I mistaking what you are saying?

Author: This will be investigated further

Reviewer 2: I think the ```VLAN=101``` can be omitted. 

If the description is clear stating that configuring PVID will also configure the VLAN, then that would be enough.

> The trunk port has the following network file.

Reviewer: ```suggestion
The trunk port is created with the following network file.
```

Author: fixed

> ```

Reviewer: If this is the trunk port, are all default VLAN ports trunk ports? and all with PVID access ports?

Author: This is an example of a trunk port, I do not fully understand enough to answer your question more than that right now. Maybe @rubensfig ?

Reviewer 2: So, all bridge ports are trunk ports, as in the port is trunked into several VLANs.

The access port is the special case where a trunk port has a single VLAN as its Native VLAN (PVID).
